src/retention-data/index.ts

```diff
@@ -0,0 +1,60 @@
+import { AzureFunction, Context } from '@azure/functions';
+
+import { getLatestRetention, getActivities } from './api';
+import { ActivityData } from './types';
+import { getDaysBetweenDates, getISODateString } from './utils';
+import { getRetentionData } from './retention';
+import { trackEvent, sendPendingData } from './analytics';
+
+const retentionData: AzureFunction = async function (context: Context, myTimer: any): Promise<void> {
+    if (myTimer.IsPastDue) {
+        context.log('Timer function is running late!');
+    }
+
+    const activitiesByDate = new Map<Date, ActivityData[]>();
+    const latestRetention = await getLatestRetention();
+    const dates: Date[] = [];
+
+    if (!latestRetention) {
+        context.log('No latest retention found, using "now" as date');
+        dates.push(new Date(getISODateString()));
+    } else {
+        const delta = getDaysBetweenDates(new Date(getISODateString()), latestRetention.date as Date);
+
+        // If delta === 0, we already have the retention data for today.
+        if (delta !== 0) {
+            const today = new Date(getISODateString());
+
+            for (let i = 0; i < delta; i++) {
+                const date = new Date(today);
+
+                date.setUTCDate(date.getUTCDate() - i);
+
+                dates.push(date);
+            }
+        }
+    }
+
+    if (!dates) {
+        context.log('No data to update.');
+
+        return;
+    }
+
+    for (const date of dates) {
```

Comment:
Is there a limitation in the number of simultaneous queries we can do? If not I think we should parallelize this, otherwise we are just making one request at a time. Also is there a way we can merge the 2 `for..of`? It seems like they are iterating more or less over the same data.

Something similar to:

```ts
const retentionDataResults = dates.map((date) => {
  // The following could be its own function for better clarity
  return getActivities(date)
    .then((activities) => { 
      const retention = getRetentionData(activities, date);
      trackEvent('f12-retention', retentionData);
    });
});

await Promise.all(retentionDataResults);
await sendPendingData();

```

Reply:
I don't think it worthy (readability wise), if everything is ok, dates will have just one item.

Reply:
Is not only readability wise, we are parallelizing the whole process instead of running sequentially. 

The following could be in a separate function for readability:

```js
return getActivities(date)
    .then((activities) => { 
      const retention = getRetentionData(activities, date);
      trackEvent('f12-retention', retentionData);
    });
```

Reply:
Ok, but what I meant is that usually, dates should be one element (except probably the first run) so it is not going to be a big improvement.

Reply:
I think you underestimate the cost of establish a new connection with a server 🙂